File: src/main/kotlin/dev/jefy/connectpro/portfolio/applicaion/command/ServiceCommand.java
package dev.jefy.connectpro.portfolio.applicaion.command;

import org.jspecify.annotations.NullMarked;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

import dev.jefy.connectpro.management.domain.vo.AwardId;
import dev.jefy.connectpro.portfolio.applicaion.dtos.FAQRequest;
import dev.jefy.connectpro.portfolio.applicaion.dtos.ServiceRequest;
import dev.jefy.connectpro.portfolio.domain.vo.FAQId;
import dev.jefy.connectpro.shared.domain.vo.ImageUrl;

/**
 * @author Jôph Yamba
 */
@NullMarked
public interface ServiceCommand {
    
    ServiceId create(ServiceRequest request);
    ServiceId update(ServiceId serviceId, ServiceRequest request);
    ServiceId setCoverImage(ServiceId serviceId, MultipartFile image) throws IOException;
    ServiceId deleteCoverImage(ServiceId serviceId) throws IOException;
    ServiceId addImage(ServiceId serviceId, MultipartFile image) throws IOException;
    ServiceId removeImage(ServiceId serviceId, ImageUrl imageUrl) throws IOException;
    ServiceId addFaq(ServiceId serviceId, FAQRequest request);
    ServiceId removeFaq(ServiceId serviceId, FAQId faqId);
    ServiceId setAward(ServiceId serviceId, AwardId  awardId);
    ServiceId removeAward(ServiceId serviceId);
    ServiceId activateService(ServiceId serviceId);
    ServiceId deactivateService(ServiceId serviceId);
    
}

File: src/main/kotlin/dev/jefy/connectpro/portfolio/application/command/ServiceCommand.kt
package dev.jefy.connectpro.portfolio.application.command

import dev.jefy.connectpro.management.domain.vo.AwardId
import dev.jefy.connectpro.portfolio.application.dtos.FAQRequest
import dev.jefy.connectpro.portfolio.application.dtos.ServiceRequest
import dev.jefy.connectpro.portfolio.domain.vo.FAQId
import dev.jefy.connectpro.portfolio.domain.vo.ServiceId
import dev.jefy.connectpro.shared.domain.vo.ImageUrl
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

interface ServiceCommand {
    fun create(request: ServiceRequest): ServiceId
    fun update(serviceId: ServiceId, request: ServiceRequest): ServiceId
    @Throws(IOException::class)
    fun setCoverImage(serviceId: ServiceId, image: MultipartFile): ServiceId
    @Throws(IOException::class)
    fun deleteCoverImage(serviceId: ServiceId): ServiceId
    @Throws(IOException::class)
    fun addImage(serviceId: ServiceId, image: MultipartFile): ServiceId
    @Throws(IOException::class)
    fun removeImage(serviceId: ServiceId, imageUrl: ImageUrl): ServiceId
    fun addFaq(serviceId: ServiceId, request: FAQRequest): ServiceId
    fun removeFaq(serviceId: ServiceId, faqId: FAQId): ServiceId
    fun setAward(serviceId: ServiceId, awardId: AwardId): ServiceId
    fun removeAward(serviceId: ServiceId): ServiceId
    fun activateService(serviceId: ServiceId): ServiceId
    fun deactivateService(serviceId: ServiceId): ServiceId
    fun delete(serviceId: ServiceId)
}